kv,sql: randomly inject retriable errors

stevendanna opened the pull request:
Occasionally, we see bugs caused by mismangement of state inside a
function that will be retried.

This is a cheap way to help us hunt down such bugs.

However, in its current form, it is too noisy to turn on even
metamorphically.

Further, it is not clear to me that this is the right shape of this
feature. An alternative would be a testing knob that allows you to
enable this on a per-test server basis; however that will be just as
noisy for the time being.

Informs https://github.com/cockroachdb/cockroach/issues/106417

Epic: none

Release note: None

stevendanna committed:
kv,sql: testing knob to randomly inject retryable errors

Occasionally, we see bugs caused by mismangement of state inside a
function that will be retried. Namely, the functions `(*kv.DB).Txn`
and `(*sql.InternalExecutor).Txn` take funcs that must be suitably
idempotent.

This is testing knob provides cheap way to help us hunt down such
bugs.

In the long run, it would be nice to turn this on by default, but we
should probably make sure TestServer runs reliably with this turned on
before flipping the default.

I also considered retrying _every_ transaction once rather than
relying on randomness. However, that would require adding more state
in kv.Txn, which I wasn't sure I really wanted to do for testing
purposes.

Informs #106417

Epic: none

Release note: None

stevendanna committed:
tenantsettingswatcher: allow duplicate settings in full update

Previously, we paniced if a duplicate setting was found in a full
update.

In the face of randomly injected transaction retries this is a problem
since if part of the rangefeed library's initial scan is retried since
the rangefeed library calls its onSpanDone callback before the
transaction is committed.

I haven't changed the rangefeed library since it isn't clear to me
that this fixed timestamp scan will actually be meaningfully retried
in practice.

However, allowing _exact_ duplicates in the settings watcher seems
unproblematic and allows us to continue experimenting with randomly
injected transaction retries.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106506)
<!-- Reviewable:end -->

stevendanna committed:
backupccl: increment id generator outside txn in test

The tests that use create-dumy-system-table currently expects the
tables at known IDs. The ID generator is non-transactional, so on each
randomly injected transaction retry, the value would be incremented,
invalidaating the assertions.

Note that the tests that use this are still a little broken, because
GenerateUniqueDescID may itself retry. We should have this command
return the name of the table and then use `let` to build assertions.

Epic: none

Release note: None

stevendanna committed:
backupccl: disable random txn retries in metadata test

It appears that CREATE STATISTICS produces duplicate statistics in the
face of transaction retries.

Epic: none

Release note: None

stevendanna committed:
backupccl: enable random transaction retries for data-driven tests

I believe we've solved enough issues to be able to turn this on for
most data-driven backup/restore tests now.

Release note: None

Epic: none

stevendanna commented:
Some of these commits are also being opened as separate PRs.

adityamaru commented on pkg/kv/sender.go:
```diff
@@ -338,6 +338,8 @@ type TxnSender interface {
 
 	// HasPerformedWrites returns true if a write has been performed.
 	HasPerformedWrites() bool
+
```
nit: comment?

ecwall commented on pkg/kv/kvclient/kvcoord/txn_coord_sender.go:
```diff
@@ -1511,6 +1523,16 @@ func (tc *TxnCoordSender) HasPerformedWrites() bool {
+	if tc.testingKnobs.EnableRandomTransactionRetryErrors {
+		return true
+	}
+	if forceTxnRetries && buildutil.CrdbTestBuild {
```
`&& buildutil.CrdbTestBuild` requires an extra step when building to debug with goland to make this work. Is this to prevent it from being used in production?

ecwall replied:
Never mind, not using this anyways.
